minicriu-client.c

```diff
@@ -307,4 +336,72 @@ int minicriu_register_new_thread(void) {
 	return 0;
 }
 
+static int mc_getmap() {
+	char line[512];
+	mc_mapscnt = 0;
+	FILE *proc_maps;
+	proc_maps = fopen("/proc/self/maps", "r");
+
+	if (!proc_maps)
+		return 1;
+
+	while (fgets(line, sizeof(line), proc_maps)) {
+		void *addr_start, *addr_end;
+		char mapname[256];
+		if (sscanf(line, "%p-%p %*s %*x %*d:%*d %*d %s",
+					&addr_start, &addr_end, mapname) < 2) {
+			fclose(proc_maps);
+			return 1;
+		}
+
+		/*
+		* there is no need to save [vsyscall] as it always
+		* maps to the same address in the kernel space
+		*/
+		if (!strncmp(mapname, "[vsyscall]", 10)) continue;
 
+		if (mc_mapscnt == MC_MAX_MAPS) {
+			fclose(proc_maps);
+			return 1;
+		}
+
+		maps[mc_mapscnt].start = addr_start;
+		maps[mc_mapscnt++].end = addr_end;
+	}
+	fclose(proc_maps);
+
+	return 0;
+}
+
+static int mc_cleanup() {
+	char line[512];
+	FILE *proc_maps = fopen("/proc/self/maps", "r");
+	void *last_map_start;
+	void *last_map_end;
+
+	// find last segment mapped in user space
+	while (fgets(line, sizeof(line), proc_maps)) {
+		void *addr_start, *addr_end;
+		char mapname[256];
+		if (sscanf(line, "%p-%p %*s %*x %*d:%*d %*d %s",
+					&addr_start, &addr_end, mapname) < 2) {
+			fclose(proc_maps);
+			return 1;
+		}
+
+		if (!strncmp(mapname, "[vsyscall]", 10)) continue;
```

Comment:
Won't we need to process `[vsyscall]` as any other mapping? If `vsyscall` is the last mapping, I think we still want it as a `last_map`, right?

Reply:
I think not, because the vsyscall page is statically allocated to the same address in every process. The location of the vsyscall page is fixed in the kernel ABI.

Reply:
Good to know, and since the code is based on this reasoning, please inlcude this info as a comment above this line.